src/PreparedWrites.cc

```diff
@@ -0,0 +1,521 @@
+ * header as well as the participant list and object.
+ * This form of the constructor is typically used for extracting
+ * information out of the log. For example to serve read requests
+ * or to perform log cleaning.
+ *
+ * \param buffer
+ *      Buffer referring to a complete PreparedOp in the log. It is the
+ *      caller's responsibility to make sure that the buffer passed in
+ *      actually contains a full object. If it does not, then behavior
+ *      is undefined.
+ * \param offset
+ *      Starting offset in the buffer where the PreparedOp begins.
+ * \param length
+ *      Total length of the PreparedOp in bytes.
+ *      A length of 0 means that the PreparedOp occupies the entire buffer
+ *      starting at offset.
+ */
+PreparedOp::PreparedOp(Buffer& buffer, uint32_t offset, uint32_t length)
+    : header(*buffer.getOffset<Header>(offset))
+    , participants(NULL)
+    , object(buffer,
+             offset + sizeof32(header) +
+               sizeof32(WireFormat::TxParticipant) * header.participantCount,
+             length - sizeof32(header) -
+               sizeof32(WireFormat::TxParticipant) * header.participantCount)
+{
+    //Fill in participants from buffer data.
+    participants = (WireFormat::TxParticipant*)buffer.getRange(
+            offset + sizeof32(header),
+            sizeof32(WireFormat::TxParticipant) * header.participantCount);
+}
+
+/**
+ * Append the serialized header, participant list, and object to the
+ * provided buffer.
+ *
+ * \param buffer
+ *      The buffer to append all data to.
+ */
+void
+PreparedOp::assembleForLog(Buffer& buffer)
+{
+    header.checksum = computeChecksum();
+    buffer.appendExternal(&header, sizeof32(Header));
+    buffer.appendExternal(participants, sizeof32(WireFormat::TxParticipant)
+                                         * header.participantCount);
+    object.assembleForLog(buffer);
+}
+
+/**
+ * Compute a checksum on the PreparedOp and determine whether or not it matches
+ * what is stored in the PreparedOp. Returns true if the checksum looks ok,
+ * otherwise returns false.
+ */
+bool
+PreparedOp::checkIntegrity()
+{
+    return computeChecksum() == header.checksum;
+}
+
+/**
+ * Compute the PreparedOp's checksum and return it.
+ */
+uint32_t
+PreparedOp::computeChecksum()
+{
+    Crc32C crc;
+    // first compute the checksum on the object header excluding the
+    // checksum field
+    crc.update(this,
+               downCast<uint32_t>(sizeof(header) -
+               sizeof(header.checksum)));
+
+    crc.update(participants,
+               sizeof32(WireFormat::TxParticipant) *
+               header.participantCount);
+
+    // TODO(seojin): Let's discuss safety of this.
+    uint32_t objectChecksum = object.computeChecksum();
+    crc.update(&objectChecksum, sizeof32(objectChecksum));
+
+    return crc.getResult();
+}
+
+PreparedOpTombstone::PreparedOpTombstone(PreparedOp& op, uint64_t segmentId)
```

Comment:
Need header comments for this method and the following one, so that the arguments are documented.